### PR TITLE
[unsat cores] Adds command to retrieve lemmas used to derive unsat

### PR DESCRIPTION
**Pull request by HanielB**
Feature is based on the SAT proof. The lemmas that are accumulated by the `ProofCnfStream` during a check-sat call are filtered according to being used or not in the SAT proof and then printed as a list.

### CODE REVIEW

**Review by ajreynol — CHANGES_REQUESTED** [review]
Can you add a regression that uses the `(get-unsat-core-lemmas)`? Else LGTM.

File: include/cvc5/cvc5.h [hunk 1]
```diff
@@ -4279,6 +4279,29 @@ class CVC5_EXPORT Solver
...
+   *     (get-unsat-core-lemmas)
+   *
+   * Requires to enable option
+   * :ref:`produce-unsat-core-lemmas <lbl-option-produce-unsat-core-lemmas>`.
+   *
+   * .. note::
+   *   This method requires computing unsat cores via the SAT proof, so other
+   *   modes of unsat core are overriden by it.
+   * \endverbatim
+   *
```
Comment (ajreynol) [inline include/cvc5/cvc5.h:4299]: Note that the feature is experimental.

File: include/cvc5/cvc5.h [hunk 2]
```diff
@@ -4279,6 +4279,29 @@ class CVC5_EXPORT Solver
...
+   * Get the lemmas used to derive unsatisfiability.
+   *
+   * SMT-LIB:
+   *
+   * \verbatim embed:rst:leading-asterisk
+   * .. code:: smtlib
+   *
+   *     (get-unsat-core-lemmas)
+   *
+   * Requires to enable option
```
Comment (ajreynol) [inline include/cvc5/cvc5.h:4292]: Is this out of date?

File: src/api/java/io/github/cvc5/Solver.java [hunk 3]
```diff
@@ -1943,6 +1943,31 @@ public Term[] getUnsatCore()
...
+   * (get-unsat-core-lemmas)
+   * }
+   * Requires to enable option {@code produce-unsat-core-lemmas}.
+   *
+   * @api.note In contrast to SMT-LIB, cvc5's API does not distinguish between
+   *           named and unnamed assertions when producing an unsatisfiable
+   *           core. Additionally, the API allows this option to be called after
+   *           a check with assumptions. A subset of those assumptions may be
+   *           included in the unsatisfiable core returned by this method.
+   *
```
Comment (ajreynol) [inline src/api/java/io/github/cvc5/Solver.java:1959]: Ditto, update the doc based on the updates to the C++ doc.

File: src/api/python/cvc5.pxi [hunk 4]
```diff
@@ -2427,6 +2427,33 @@ cdef class Solver:
...
+
+                (get-unsat-core-lemmas)
+
+            Requires to enable option
+            :ref:`produce-unsat-core-lemmas <lbl-option-produce-unsat-core-lemmas>`.
+
+            .. note::
+              This method requires computing unsat cores via the SAT proof, so other
+              modes of unsat core are overriden by it.
+
```
Comment (ajreynol) [inline src/api/python/cvc5.pxi:2446]: Ditto, update the doc based on the updates to the C++ doc.

File: src/smt/unsat_core_manager.cpp [hunk 5]
```diff
@@ -46,7 +46,7 @@ void UnsatCoreManager::getUnsatCore(std::shared_ptr<ProofNode> pfn,
   Trace("unsat-core") << "UCManager::getUnsatCore: free assumptions: "
                       << fassumps << "\n";
   const context::CDList<Node>& al = as.getAssertionList();
-  std::unordered_set<Node> coreSet;
+  std::set<Node> coreSet;
```
Comment (ajreynol) [inline src/smt/unsat_core_manager.cpp:49]: Unordered set is faster, we should continue to use std::unordered_set here.

**Review by ajreynol — CHANGES_REQUESTED** [review]
I believe unit tests are missing for C++.

File: test/regress/cli/regress0/cores/unsat-core-lemmas.smt2 [hunk 6]
```diff
@@ -0,0 +1,37 @@
+; COMMAND-LINE: --simplification=none --unsat-cores-mode=sat-proof
+; EXPECT: unsat
+; EXPECT: (
+; EXPECT: a0
```
Comment (ajreynol) [inline test/regress/cli/regress0/cores/unsat-core-lemmas.smt2:4]: I would omit the names and just expect 
```
(
)
```

Comment (ajreynol) [reply test/regress/cli/regress0/cores/unsat-core-lemmas.smt2:4]: Or just conjoin all assertions in the unsat core and have a single name expected.

Comment (HanielB) [reply test/regress/cli/regress0/cores/unsat-core-lemmas.smt2:4]: I do not understand.

File: src/smt/unsat_core_manager.cpp (hunk 5 above)
Comment (HanielB) [reply src/smt/unsat_core_manager.cpp:50]: Using `set` ensures the order is the same across different platforms.

File: include/cvc5/cvc5.h (hunk 2 above)
Comment (HanielB) [reply include/cvc5/cvc5.h:4396]: It was. Will fix.

File: test/regress/cli/regress0/cores/unsat-core-lemmas.smt2 (hunk 6 above)
Comment (ajreynol) [reply test/regress/cli/regress0/cores/unsat-core-lemmas.smt2:4]: This benchmark output is sensitive to the order of names in the unsat core, I would design the test to not rely on this.

**Comment by HanielB** [comment]
Hopefully it's good to go now, PTAL

**Review by ajreynol — APPROVED** [review]

**Pull request merged**